### PR TITLE
connect Kernel and Linear machine to a signal handler

### LINKED ISSUE #4283: Subscription to a handler in derivatives of CMachine
Currently only [CMachine::train](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/Machine.cpp#L60) is subscribing to a handler.
Due to this we are not able to write ```on_pause_impl()``` of our own in derivatives of for example [CLinearMachine](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/LinearMachine.cpp#L184).

### PR DESCRIPTION
**Pull request by shubham808**
This should fix #4283

### CODE REVIEW

File: src/shogun/machine/LinearMachine.cpp [hunk 1]
```diff
@@ -180,10 +181,12 @@ bool CLinearMachine::train(CFeatures* data)
...
     }
+	auto sub = connect_to_signal_handler();
+	bool result = train_machine(data);
+	sub.unsubscribe();
+	reset_computation_variables();
 
-    bool result = train_machine(data);
-
-    if(m_compute_bias)
+	if(m_compute_bias)
```
Comment (geektoni) [inline src/shogun/machine/LinearMachine.cpp]: Pay attention to the `if` indentation here :) It seems to be wrong from Github.

**Comment by vigsterkr** [comment]
@shubham808 could you rebase your branch over latest develop to get this PR properly tested! thnx!

**Comment by shubham808** [comment]
@vigsterkr sure i am on it :)

File: src/shogun/machine/KernelMachine.cpp [hunk 2]
```diff
@@ -452,8 +452,10 @@ bool CKernelMachine::train_locked(SGVector<index_t> indices)
 
 	/* dont do train because model should not be stored (no acutal features)
 	 * and train does data_unlock */
+	auto sub = connect_to_signal_handler();
```
Comment (karlnapf) [inline src/shogun/machine/KernelMachine.cpp]: I wonder whether we shouldnt do this in the base class? I.e. `CMachine::train` ?
This way we need to register every machine specialization. Or do I see this wrong?

Comment (geektoni) [reply src/shogun/machine/KernelMachine.cpp]: It is already implemented into CMachine:train. However, there are some classes which implement their own version of the train method and therefore they do not have the connection with the signal handler.

Comment (karlnapf) [reply src/shogun/machine/KernelMachine.cpp]: ok, as discussed, let's try to move this into the base class and for the spezializations that overload the train method can just call the base class (e.g. in CLinearMachine)

File: src/shogun/machine/KernelMachine.h [hunk 3]
```diff
@@ -227,12 +227,9 @@ class CKernelMachine : public CMachine
 		virtual float64_t apply_one(int32_t num);
 
 #ifndef SWIG // SWIG should skip this part
-		/** Trains a locked machine on a set of indices. Error if machine is
-		 * not locked
-		 *
-		 * @param indices index vector (of locked features) that is used for training
-		 * @return whether training was successful
-		 */
+		     /**
```
Comment (karlnapf) [inline src/shogun/machine/KernelMachine.h]: ?

File: src/shogun/machine/Machine.cpp [hunk 4]
```diff
@@ -67,6 +67,23 @@ bool CMachine::train(CFeatures* data)
 	return result;
 }
 
+bool CMachine::train_locked()
+{
+	/*train machine without any actual features(data is locked)*/
+	SG_DEBUG("entering %s::train_locked()\n", get_name())
```
Comment (karlnapf) [inline src/shogun/machine/Machine.cpp]: these kind of messages can go

File: src/shogun/machine/Machine.cpp [hunk 5]
```diff
@@ -67,6 +67,23 @@ bool CMachine::train(CFeatures* data)
...
+
+	if (!is_data_locked())
+		SG_ERROR("%s::train_locked() call data_lock() before!\n", get_name())
+
+	auto sub = connect_to_signal_handler();
+	bool result = train_machine();
```
Comment (karlnapf) [inline src/shogun/machine/Machine.cpp]: I am not sure I understand the purpose of this method?

File: src/shogun/machine/Machine.h [hunk 6]
```diff
@@ -162,6 +162,17 @@ class CMachine : public CStoppableSGObject
 		 */
 		virtual bool train(CFeatures* data=NULL);
 
+#ifndef SWIG // SWIG should skip this part
+		/** Trains a locked machine on a set of indices. Error if machine is
+		 * not locked
+		 *
+		 * @param indices index vector (of locked features) that is used for
+		 * training
```
Comment (karlnapf) [inline src/shogun/machine/Machine.h]: the docs don't match the method

**Comment by karlnapf** [comment]
style checker is not happy: https://travis-ci.org/shogun-toolbox/shogun/jobs/382682896#L726

**Pull request merged**